src/ompl/geometric/planners/experience/src/LRR.cpp

```diff
@@ -0,0 +1,610 @@
+        repairPlanner_ = repair_planner;
+    }
+
+    // Setup the problem definition for the repair planner
+    repairProblemDef_->setOptimizationObjective(pdef_->getOptimizationObjective()); // copy primary problem def
+
+    // Setup repair planner
+    repairPlanner_->setProblemDefinition(repairProblemDef_);
+    if (!repairPlanner_->isSetup())
+        repairPlanner_->setup();
+}
+
+void ompl::geometric::LRR::freeMemory(void)
+{
+}
+
+ompl::base::PlannerStatus ompl::geometric::LRR::solve(const base::PlannerTerminationCondition &ptc)
+{
+    bool solved = false;
+    bool approximate = false;
+    double approxdif = std::numeric_limits<double>::infinity();
+
+    // Check if the database is empty
+    if (!experienceDB_->getExperiencesCount())
+    {
+        OMPL_INFORM("Experience database is empty so unable to run LRR algorithm.");
+
+        return base::PlannerStatus::CRASH;
+    }
+
+    // Get a single start state TODO: more than one
+    const base::State *startState = pis_.nextStart();
+
+    // Get a single goal state TODO: more than one
+    base::Goal *goal   = pdef_->getGoal().get();
+
+    // Check that we have the correct type of goal
+    if (!goal)
+    {
+        OMPL_ERROR("Goal cannot be converted into a goal state");
+        return base::PlannerStatus::UNRECOGNIZED_GOAL_TYPE;
+    }
+
+    const base::State *goalState = pis_.nextGoal(ptc);
+
+    // Search for previous solution in database
+    nearestPaths_ = experienceDB_->findNearestStartGoal(nearestK_, startState, goalState);
+
+    // Check if there are any solutions
+    if (nearestPaths_.empty())
+    {
+        OMPL_INFORM("No similar path founds in nearest neighbor tree, unable to retrieve repair");
+        return base::PlannerStatus::TIMEOUT; // The planner failed to find a solution
+    }
+
+    ompl::base::PlannerDataPtr chosenPath;
+
+    // Filter top n paths to 1
+    // TODO Rather than selecting 1 best path, you could also spawn n (n<=k) threads and repair the top n paths.
+    if (!findBestPath(startState, goalState, chosenPath))
+    {
+        return base::PlannerStatus::CRASH;
+    }
+
+    // All saved trajectories should be at least 2 states long
+    assert(chosenPath->numVertices() >= 2);
+
+    // Convert chosen PlanningData experience to an actual path
+    ompl::geometric::PathGeometric *primaryPath = new PathGeometric(si_);
+    // Add start
+    primaryPath->append(startState);
+    // Add old states
+    for (std::size_t i = 0; i < chosenPath->numVertices(); ++i)
+    {
+        primaryPath->append(chosenPath->getVertex(i).getState());
+    }
+    // Add goal
+    primaryPath->append(goalState);
+
+    // All save trajectories should be at least 2 states long, and then we append the start and goal states
+    assert(primaryPath->getStateCount() >= 4);
+
+    // Repair chosen path
+    if (!repairPath(*primaryPath, ptc))
+    {
+        OMPL_WARN("repairPath failed");
```

Comment:
I actually changed this the past month while using Lightning with Amazon - it now is

```
OMPL_INFORM("repairPath failed or aborted");
```

Because often the function ends simply because planning from scratch was faster